README.md
# two-firewatch

Based on https://github.com/rakr/vim-two-firewatch

This is still really early in development. I hope to eventually have it more stable.

## Running

`npm run build` will build the theme.
`npm run dev` will build the theme and watch for changes.

## Current Plan

- [ ] Stabilize around the current colors
- [ ] Make a minimal and a full version
  - I prefer minimal coloring, but I also want a faithful recreation of the original theme
- [ ] Audit the light theme
- [ ] Flesh out TS/JS support
- [ ] Flesh out jupyter support


CHANGELOG.md
# Change Log

All notable changes to the "two-firewatch" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release


generate_theme.ts
const { writeFileSync } = require("fs");

type Colors = Record<string, string>;

const lightColors: Colors = {
  uno_1: "#2d2006",
  uno_2: "#896724",
  uno_3: "#B29762",
  uno_4: "#B6ad9a",

  duo_1: "#065289",
  duo_2: "#718ecd",
  duo_3: "#aeb3b7",

  syntax_color_renamed: "#33a0ff",
  syntax_color_added: "#43d08a",
  syntax_color_modified: "#e0c285",
  syntax_color_removed: "#e05252",

  syntax_fg: "#896724",
  syntax_bg: "#FAF8F5",
  syntax_accent: "#447EBB",
  syntax_gutter: "#EAE1D2",
  syntax_selection: "#E5DDCB",
  syntax_fold_bg: "#d1cec7",
  syntax_cursor_line: "#F3EFE7",
};

/*
  *   let g:terminal_color_0 = "#282c34"
  let g:terminal_color_8 = "#282c34"
  let g:terminal_color_1 = "#e06c75"
  let g:terminal_color_9 = "#e06c75"
  let g:terminal_color_2 = "#98c379"
  let g:terminal_color_10 = "#98c379"
  let g:terminal_color_3 = "#e5c07b"
  let g:terminal_color_11 = "#e5c07b"
  let g:terminal_color_4 = "#61afef"
  let g:terminal_color_12 = "#61afef"
  let g:terminal_color_5 = "#c678dd"
  let g:terminal_color_13 = "#c678dd"
  let g:terminal_color_6 = "#56b6c2"
  let g:terminal_color_14 = "#56b6c2"
  let g:terminal_color_7 = "#dcdfe4"
  let g:terminal_color_15 = "#dcdfe4"
*/
const darkColors: Colors = {
  uno_1: "#d6e9ff",
  uno_2: "#abb2bf",
  uno_3: "#6e88a6",
  uno_4: "#55606d",

  duo_1: "#c8ae9d",
  duo_2: "#e06c75",
  duo_3: "#dd672c",

  syntax_color_renamed: "#33a0ff",
  syntax_color_added: "#43d08a",
  syntax_color_modified: "#e0c285",
  syntax_color_removed: "#e05252",

  syntax_fg: "#abb2bf",
  syntax_bg: "#282c34",
  syntax_accent: "#56b6c2",
  syntax_gutter: "#636d83",
  syntax_selection: "#3e4452",
  syntax_fold_bg: "#5c6370",
  syntax_cursor_line: "#2c323c",
};

type Theme = {
  name: string;
  colors: {
    [key: string]: string;
  };
  tokenColors: {
    name: string;
    scope: string[];
    settings: {
      foreground?: string;
      fontStyle?: string;
    };
  }[];
};

const makeTheme = (c: Colors): Theme => {
  const foreground = c.syntax_fg;

  return {
    name: "Two Firewatch",
    colors: {
      "editor.background": c.syntax_bg,
      "editor.foreground": foreground,
      "activityBarBadge.background": c.syntax_bg,
      "sideBarTitle.foreground": c.syntax_fg,
      "sideBar.background": c.syntax_bg,
      "sideBar.foreground": c.syntax_fg,
      "sideBarSectionHeader.background": c.syntax_bg,
      "sideBarSectionHeader.foreground": c.syntax_fg,
      "editorGroupHeader.tabsBackground": c.syntax_bg,
      "editorGroupHeader.tabsBorder": c.syntax_bg,
      "tab.activeBackground": c.syntax_bg,
      "tab.activeForeground": c.syntax_fg,
      "tab.inactiveBackground": c.syntax_bg,
      "tab.inactiveForeground": c.syntax_fg,
      "tab.border": c.syntax_bg,
      "editorLineNumber.foreground": c.syntax_gutter,
      "editorLineNumber.activeForeground": c.syntax_gutter,
      "editorCursor.foreground": c.syntax_fg,
      "editor.selectionBackground": c.syntax_selection,
      "editor.selectionHighlightBackground": c.syntax_selection,
      "editor.wordHighlightBackground": c.syntax_selection,
      "editor.wordHighlightStrongBackground": c.syntax_selection,
      "editor.findMatchBackground": c.syntax_selection,
      "editor.findMatchHighlightBackground": c.syntax_selection,
      "editor.lineHighlightBackground": c.syntax_cursor_line,
      "editorWhitespace.foreground": c.syntax_gutter,
      "editorIndentGuide.background": c.syntax_gutter,
      "editorIndentGuide.activeBackground": c.syntax_gutter,
      "editorRuler.foreground": c.syntax_gutter,
      "editorCodeLens.foreground": c.syntax_gutter,
      "editorBracketMatch.background": c.syntax_selection,
      "editorBracketMatch.border": c.syntax_selection,
    },
    tokenColors: [
      {
        name: "Punctuation",
        scope: ["punctuation", "expression"],
        settings: {
          foreground: c.uno_4,
        },
      },
      {
        name: "Comment",
        scope: ["comment", "punctuation.definition.comment"],
        settings: {
          fontStyle: "italic",
          foreground: c.uno_4,
        },
      },
      {
        name: "Variables",
        scope: ["variable", "string constant.other.placeholder"],
        settings: {
          foreground: c.duo_2,
        },
      },
      {
        name: "Colors",
        scope: ["constant.other.color"],
        settings: {
          foreground,
        },
      },
      {
        name: "Invalid",
        scope: ["invalid", "invalid.illegal"],
        settings: {
          foreground,
        },
      },
      {
        name: "Keyword, Storage",
        scope: [
          "keyword",
          "storage.type.js",
          "storage.type.function.js",
          "variable.other.flowtype.js",
          "storage.type.extends.js",
          "storage.type.function.js",
          "storage.type.class.js",
          "meta.class.tsx",
        ],
        settings: {
          foreground: c.uno_4,
        },
      },
      {
        name: "Operator, Misc",
        scope: [
          "keyword.control",
          "constant.other.color",
          "punctuation",
          "meta.tag",
          "punctuation.definition.tag",
          "punctuation.separator.inheritance.php",
          "punctuation.definition.tag.html",
          "punctuation.definition.tag.begin.html",
          "punctuation.definition.tag.end.html",
          "meta.brace.square.tsx",
          "meta.type.tuple.tsx",
          "punctuation.section.embedded",
          "keyword.other.template",
          "keyword.other.substitution",
        ],
        settings: {
          foreground: c.uno_4,
        },
      },
      {
        name: "Tag",
        scope: [
          "entity.name.tag",
          "meta.tag.sgml",
          "markup.deleted.git_gutter",
        ],
        settings: {
          foreground,
        },
      },
      {
        name: "Function, Special Method",
        scope: [
          "entity.name.function",
          "meta.function-call",
          "variable.function",
          "support.function",
          "keyword.other.special-method",
        ],
        settings: {
          foreground,
        },
      },
      {
        name: "Block Level Variables",
        scope: ["meta.block variable.other"],
        settings: {
          foreground: c.duo_2,
        },
      },
      {
        name: "Other Variable, String Link",
        scope: ["support.other.variable", "string.other.link"],
        settings: {
          foreground: c.duo_2,
        },
      },
      {
        name: "Function Argument, Tag Attribute, Embedded",
        scope: [
          "variable.parameter",
          "keyword.other.unit",
          "keyword.other",
          "meta.object-literal.key.tsx",
          "meta.object.member.tsx",
        ],
        settings: {
          foreground,
        },
      },
      {
        name: "Types",
        scope: ["entity.name.type.tsx", "meta.type.annotation.tsx"],
        settings: {
          foreground: c.duo_3,
        },
      },
      {
        name: "String, Symbols, Inherited Class, Markup Heading, Number, Constant",
        scope: [
          "string",
          "constant.other.symbol",
          "constant.other.key",
          "entity.other.inherited-class",
          "markup.heading",
          "markup.inserted.git_gutter",
          "meta.group.braces.curly constant.other.object.key.js string.unquoted.label.js",
          "constant.numeric",
          "constant.language",
          "support.constant",
          "constant.character",
          "constant.escape",
        ],
        settings: {
          foreground: c.duo_1,
        },
      },
      {
        name: "Class, Support",
        scope: [
          "entity.name",
          "support.type",
          "support.class",
          "support.other.namespace.use.php",
          "meta.use.php",
          "support.other.namespace.php",
          "markup.changed.git_gutter",
          "support.type.sys-types",
        ],
        settings: {
          foreground: c.duo_1,
        },
      },
      {
        name: "Entity Types",
        scope: ["support.type"],
        settings: {
          foreground: c.duo_1,
        },
      },
      {
        name: "Sub-methods",
        scope: [
          "entity.name.module.js",
          "variable.import.parameter.js",
          "variable.other.class.js",
        ],
        settings: {
          foreground: c.uno_1,
        },
      },
      {
        name: "Language methods",
        scope: ["variable.language"],
        settings: {
          fontStyle: "italic",
          foreground: c.uno_4,
        },
      },
      {
        name: "entity.name.method.js",
        scope: ["entity.name.method.js"],
        settings: {
          fontStyle: "italic",
          foreground: c.uno_1,
        },
      },
      {
        name: "meta.method.js",
        scope: [
          "meta.class-method.js entity.name.function.js",
          "variable.function.constructor",
        ],
        settings: {
          foreground: c.uno_1,
        },
      },
      {
        name: "Attributes",
        scope: ["entity.other.attribute-name"],
        settings: {
          foreground: c.uno_1,
        },
      },
      {
        name: "HTML Attributes",
        scope: [
          "text.html.basic entity.other.attribute-name.html",
          "text.html.basic entity.other.attribute-name",
        ],
        settings: {
          fontStyle: "italic",
          foreground: c.uno_1,
        },
      },
      {
        name: "URL",
        scope: ["*url*", "*link*", "*uri*"],
        settings: {
          fontStyle: "underline",
        },
      },
      {
        name: "Decorators",
        scope: [
          "tag.decorator.js entity.name.tag.js",
          "tag.decorator.js punctuation.definition.tag.js",
        ],
        settings: {
          fontStyle: "italic",
          foreground: c.duo_1,
        },
      },
      {
        name: "ES7 Bind Operator",
        scope: [
          "source.js constant.other.object.key.js string.unquoted.label.js",
        ],
        settings: {
          fontStyle: "italic",
          foreground: c.duo_1,
        },
      },
    ],
  };
};

// write theme to json file
writeFileSync(
  "themes/Two Firewatch-color-theme-dark.json",
  JSON.stringify(makeTheme(darkColors), null, 2)
);
writeFileSync(
  "themes/Two Firewatch-color-theme-light.json",
  JSON.stringify(makeTheme(lightColors), null, 2)
);


src/build.ts
import { writeFileSync } from "fs";
import { join } from "path";

const toRGB = (hex: string) => {
  const red = parseInt(hex[1] + hex[2], 16);
  const green = parseInt(hex[3] + hex[4], 16);
  const blue = parseInt(hex[5] + hex[6], 16);
  return { red, green, blue };
};

const opacity = (color: string, bg: string, opacity: number) => {
  const foreground = toRGB(color);
  const background = toRGB(bg);

  const red = Math.round(
    foreground.red * opacity + background.red * (1 - opacity)
  );
  const green = Math.round(
    foreground.green * opacity + background.green * (1 - opacity)
  );
  const blue = Math.round(
    foreground.blue * opacity + background.blue * (1 - opacity)
  );

  const toHexString = (num: number) => num.toString(16).padStart(2, "0");
  return "#" + toHexString(red) + toHexString(green) + toHexString(blue);
};

type Colors = {
  uno1: string;
  uno2: string;
  uno3: string;
  uno4: string;

  duo1: string;
  duo2: string;
  duo3: string;

  syntaxColorRenamed: string;
  syntaxColorAdded: string;
  syntaxColorModified: string;
  syntaxColorRemoved: string;

  syntaxFg: string;
  syntaxBg: string;
  panelBg: string;
  syntaxAccent: string;
  syntaxGutter: string;
  syntaxSelection: string;
  syntaxFoldBg: string;
  syntaxCursorLine: string;

  background1: string;
  background2: string;
  background3: string;
  background4: string;

  terminalColor0: string;
  terminalColor1: string;
  terminalColor2: string;
  terminalColor3: string;
  terminalColor4: string;
  terminalColor5: string;
  terminalColor6: string;
  terminalColor7: string;
  terminalColor8: string;
  terminalColor9: string;
  terminalColor10: string;
  terminalColor11: string;
  terminalColor12: string;
  terminalColor13: string;
  terminalColor14: string;
  terminalColor15: string;
};

const lightColors: Colors = {
  uno1: "#2d2006",
  uno2: "#896724",
  uno3: "#B29762",
  uno4: "#B6ad9a",

  duo1: "#065289",
  duo2: "#718ecd",
  duo3: "#aeb3b7",

  syntaxColorRenamed: "#33a0ff",
  syntaxColorAdded: "#43d08a",
  syntaxColorModified: "#e0c285",
  syntaxColorRemoved: "#e05252",

  syntaxFg: "#896724",
  syntaxBg: "#FAF8F5",
  panelBg: "#F3EFE7",
  syntaxAccent: "#447EBB",
  syntaxGutter: "#EAE1D2",
  syntaxSelection: "#E5DDCB",
  syntaxFoldBg: "#d1cec7",
  syntaxCursorLine: "#F3EFE7",

  background1: "#FAF8F5",
  background2: opacity("#FAF8F5", "#ffffff", 0.5),
  background3: opacity("#FAF8F5", "#ffffff", 0.25),
  background4: opacity("#FAF8F5", "#ffffff", 0.25),

  terminalColor0: "#282c34",
  terminalColor1: "#e06c75",
  terminalColor2: "#98c379",
  terminalColor3: "#e5c07b",
  terminalColor4: "#61afef",
  terminalColor5: "#c678dd",
  terminalColor6: "#56b6c2",
  terminalColor7: "#dcdfe4",
  terminalColor8: "#282c34",
  terminalColor9: "#e06c75",
  terminalColor10: "#98c379",
  terminalColor11: "#e5c07b",
  terminalColor12: "#61afef",
  terminalColor13: "#c678dd",
  terminalColor14: "#56b6c2",
  terminalColor15: "#dcdfe4",
};

const darkColors: Colors = {
  uno1: "#d6e9ff",
  uno2: "#abb2bf",
  uno3: "#6e88a6",
  uno4: "#55606d",

  duo1: "#c8ae9d",
  duo2: "#e06c75",
  duo3: "#dd672c",

  syntaxColorRenamed: "#33a0ff",
  syntaxColorAdded: "#43d08a",
  syntaxColorModified: "#e0c285",
  syntaxColorRemoved: "#e05252",

  syntaxFg: "#abb2bf",
  syntaxBg: "#282c34",
  panelBg: "#0C1119",
  syntaxAccent: "#56b6c2",
  syntaxGutter: "#636d83",
  syntaxSelection: "#3e4452",
  syntaxFoldBg: "#5c6370",
  syntaxCursorLine: "#2c323c",

  background1: "#282c34",
  background2: opacity("#282c34", "#000000", 0.75),
  background3: opacity("#282c34", "#000000", 0.5),
  background4: opacity("#282c34", "#000000", 0.25),

  terminalColor0: "#282c34",
  terminalColor1: "#e06c75",
  terminalColor2: "#98c379",
  terminalColor3: "#e5c07b",
  terminalColor4: "#61afef",
  terminalColor5: "#c678dd",
  terminalColor6: "#56b6c2",
  terminalColor7: "#dcdfe4",
  terminalColor8: "#282c34",
  terminalColor9: "#e06c75",
  terminalColor10: "#98c379",
  terminalColor11: "#e5c07b",
  terminalColor12: "#61afef",
  terminalColor13: "#c678dd",
  terminalColor14: "#56b6c2",
  terminalColor15: "#dcdfe4",
};

type Theme = {
  name: string;
  colors: {
    [key: string]: string;
  };
  tokenColors: {
    name: string;
    scope: string[];
    settings: {
      foreground?: string;
      background?: string;
      fontStyle?: string;
    };
  }[];
};

const makeTheme = (c: Colors): Theme => {
  const foreground = c.syntaxFg;
  const o = (fg: string, op: number) => opacity(fg, c.syntaxBg, op);

  const theme: Theme = {
    name: "Two Firewatch",
    colors: {
      // Reference doc https://code.visualstudio.com/api/references/theme-color
      "editor.background": c.syntaxBg,
      "editor.foreground": foreground,
      "editorGroup.border": c.background2,
      "editorGroup.dropBackground": "#ffffff00",
      "editorGroup.emptyBackground": c.background2,
      "editor.renderLineHighlight": c.syntaxCursorLine,

      "selection.background": c.syntaxSelection,

      "activityBarBadge.background": c.syntaxBg,
      "sideBarTitle.foreground": c.syntaxFg,
      "sideBar.background": c.background3,
      "sideBar.foreground": c.syntaxFg,
      "sidebar.border": c.background3,
      "sideBarSectionHeader.background": c.background3,
      "sideBarSectionHeader.foreground": c.syntaxFg,
      "sidebar.selectionBackground": c.syntaxSelection,
      "sidebar.selectionForeground": c.syntaxFg,
      "editorGroupHeader.tabsBackground": c.background3,
      "editorGroupHeader.tabsBorder": c.syntaxBg,
      "tab.activeBackground": c.background3,
      "tab.activeForeground": c.syntaxFg,
      "tab.inactiveBackground": c.background3,
      "tab.inactiveForeground": c.syntaxFg,
      "tab.border": c.syntaxBg,
      "editorLineNumber.foreground": c.syntaxGutter,
      "editorLineNumber.activeForeground": c.syntaxGutter,
      "editorCursor.foreground": c.syntaxFg,
      "editor.selectionBackground": c.syntaxSelection,
      "editor.selectionHighlightBackground": c.syntaxSelection,
      "editor.wordHighlightBackground": c.syntaxSelection,
      "editor.wordHighlightStrongBackground": c.syntaxSelection,
      "editor.findMatchBackground": c.syntaxSelection,
      "editor.findMatchHighlightBackground": c.syntaxSelection,
      "editor.lineHighlightBackground": c.syntaxCursorLine,
      "editorWhitespace.foreground": c.syntaxGutter,
      "editorIndentGuide.background": c.background4,
      "editorIndentGuide.activeBackground": c.background4,
      "editorRuler.foreground": c.syntaxGutter,
      "editorCodeLens.foreground": c.syntaxGutter,
      "editorBracketMatch.background": c.syntaxSelection,

      // Git
      "gitDecoration.addedResourceForeground": c.syntaxColorAdded,
      "gitDecoration.modifiedResourceForeground": c.syntaxColorModified,
      "gitDecoration.deletedResourceForeground": c.syntaxColorRemoved,
      "gitDecoration.untrackedResourceForeground": c.syntaxColorAdded,
      "diffEditor.insertedTextBackground": o(c.syntaxColorAdded, 0.1),
      "diffEditor.removedTextBackground": o(c.syntaxColorRemoved, 0.1),

      // Removing shadow
      "scollbar.shadow": "#ffffff00",
      "widget.shadow": "#ffffff00",

      // Panel / Terminal
      "terminal.foreground": c.syntaxFg,
      "terminal.background": c.background2,
      "panel.background": c.background2,
      "panel.border": c.background2,
      "panelTitle.activeBorder": c.syntaxAccent,
      "panelTitle.activeForeground": c.syntaxFg,
      "panelTitle.inactiveForeground": c.syntaxGutter,
      "statusBar.background": c.background2,
      "statusBar.foreground": c.syntaxFg,

      // Prompt
      "prompt.background": c.background2,
      "prompt.foreground": c.syntaxFg,
      "prompt.border": c.background4,
      "prompt.inputBackground": c.background1,
      "prompt.inputForeground": c.syntaxFg,
      "prompt.inputBorder": c.background4,

      "terminal.ansiBlack": c.terminalColor0,
      "terminal.ansiRed": c.terminalColor1,
      "terminal.ansiGreen": c.terminalColor2,
      "terminal.ansiYellow": c.terminalColor3,
      "terminal.ansiBlue": c.terminalColor4,
      "terminal.ansiMagenta": c.terminalColor5,
      "terminal.ansiCyan": c.terminalColor6,
      "terminal.ansiWhite": c.terminalColor7,
      "terminal.ansiBrightBlack": c.terminalColor8,
      "terminal.ansiBrightRed": c.terminalColor9,
      "terminal.ansiBrightGreen": c.terminalColor10,
      "terminal.ansiBrightYellow": c.terminalColor11,
      "terminal.ansiBrightBlue": c.terminalColor12,
      "terminal.ansiBrightMagenta": c.terminalColor13,
      "terminal.ansiBrightCyan": c.terminalColor14,
      "terminal.ansiBrightWhite": c.terminalColor15,
    },
    tokenColors: [
      {
        name: "Punctuation",
        scope: ["punctuation", "expression"],
        settings: {
          foreground: c.uno4,
        },
      },
      {
        name: "Comment",
        scope: ["comment", "punctuation.definition.comment"],
        settings: {
          fontStyle: "italic",
          foreground: c.uno4,
        },
      },
      {
        name: "Variables",
        scope: ["variable", "string constant.other.placeholder"],
        settings: {
          foreground: c.duo2,
        },
      },
      {
        name: "Colors",
        scope: ["constant.other.color"],
        settings: {
          foreground,
        },
      },
      {
        name: "Invalid",
        scope: ["invalid", "invalid.illegal"],
        settings: {
          foreground,
        },
      },
      {
        name: "Keyword, Storage",
        scope: [
          "keyword",
          "storage.type.js",
          "storage.type.function.js",
          "variable.other.flowtype.js",
          "storage.type.extends.js",
          "storage.type.function.js",
          "storage.type.class.js",
          "meta.class.tsx",
        ],
        settings: {
          foreground: c.uno4,
        },
      },
      {
        name: "Operator, Misc",
        scope: [
          "constant.other.color",
          "punctuation",
          "meta.tag",
          "punctuation.definition.tag",
          "punctuation.separator.inheritance.php",
          "punctuation.definition.tag.html",
          "punctuation.definition.tag.begin.html",
          "punctuation.definition.tag.end.html",
          "meta.brace.square.tsx",
          "meta.type.tuple.tsx",
          "punctuation.section.embedded",
          "keyword.other.template",
          "keyword.other.substitution",
        ],
        settings: {
          foreground: c.uno4,
        },
      },
      {
        name: "Tag",
        scope: [
          "entity.name.tag",
          "meta.tag.sgml",
          "markup.deleted.git_gutter",
        ],
        settings: {
          foreground,
        },
      },
      {
        name: "Function, Special Method",
        scope: [
          "entity.name.function",
          "meta.function-call",
          "variable.function",
          "support.function",
          "keyword.other.special-method",
        ],
        settings: {
          foreground,
        },
      },
      {
        name: "Block Level Variables",
        scope: ["meta.block variable.other"],
        settings: {
          foreground: c.duo2,
        },
      },
      {
        name: "Other Variable, String Link",
        scope: ["support.other.variable", "string.other.link"],
        settings: {
          foreground: c.duo2,
        },
      },
      {
        name: "Function Argument, Tag Attribute, Embedded",
        scope: [
          "variable.parameter",
          "keyword.other.unit",
          "keyword.other",
          "meta.object-literal.key.tsx",
          "meta.object.member.tsx",
        ],
        settings: {
          foreground,
        },
      },
      {
        name: "String, Symbols, Inherited Class, Markup Heading, Number, Constant",
        scope: [
          "string",
          "constant.other.symbol",
          "constant.other.key",
          "entity.other.inherited-class",
          "markup.heading",
          "markup.inserted.git_gutter",
          "meta.group.braces.curly constant.other.object.key.js string.unquoted.label.js",
          "constant.numeric",
          "constant.language",
          "support.constant",
          "constant.character",
          "constant.escape",
        ],
        settings: {
          foreground: c.duo1,
        },
      },
      {
        name: "Class, Support",
        scope: [
          "entity.name",
          "support.type",
          "support.class",
          "support.other.namespace.use.php",
          "meta.use.php",
          "support.other.namespace.php",
          "markup.changed.git_gutter",
          "support.type.sys-types",
        ],
        settings: {
          foreground: c.duo1,
        },
      },
      {
        name: "Entity Types",
        scope: ["support.type"],
        settings: {
          foreground: c.duo1,
        },
      },
      {
        name: "Sub-methods",
        scope: [
          "entity.name.module.js",
          "variable.import.parameter.js",
          "variable.other.class.js",
        ],
        settings: {
          foreground: c.uno1,
        },
      },
      {
        name: "Language methods",
        scope: ["variable.language"],
        settings: {
          fontStyle: "italic",
          foreground: c.uno4,
        },
      },
      {
        name: "entity.name.method.js",
        scope: ["entity.name.method.js"],
        settings: {
          fontStyle: "italic",
          foreground: c.uno1,
        },
      },
      {
        name: "meta.method.js",
        scope: [
          "meta.class-method.js entity.name.function.js",
          "variable.function.constructor",
        ],
        settings: {
          foreground: c.uno1,
        },
      },
      {
        name: "Attributes",
        scope: ["entity.other.attribute-name"],
        settings: {
          foreground: c.uno1,
        },
      },
      {
        name: "HTML Attributes",
        scope: [
          "text.html.basic entity.other.attribute-name.html",
          "text.html.basic entity.other.attribute-name",
        ],
        settings: {
          fontStyle: "italic",
          foreground: c.uno1,
        },
      },
      {
        name: "URL",
        scope: ["*url*", "*link*", "*uri*"],
        settings: {
          fontStyle: "underline",
        },
      },
      {
        name: "Decorators",
        scope: [
          "tag.decorator.js entity.name.tag.js",
          "tag.decorator.js punctuation.definition.tag.js",
        ],
        settings: {
          fontStyle: "italic",
          foreground: c.duo1,
        },
      },
      {
        name: "ES7 Bind Operator",
        scope: [
          "source.js constant.other.object.key.js string.unquoted.label.js",
        ],
        settings: {
          fontStyle: "italic",
          foreground: c.duo1,
        },
      },
    ],
  };

  // Adding typescript specific styling
  theme.tokenColors = theme.tokenColors.concat([
    {
      name: "TSX focused",
      scope: ["meta.jsx.children.tsx", "entity.name.tag.tsx"],
      settings: {
        foreground: c.uno1,
      },
    },
    {
      name: "TSX dimmed",
      scope: ["entity.other.attribute-name.tsx"],
      settings: {
        foreground: c.uno2,
      },
    },
    {
      name: "Typescript Punctuation",
      scope: ["meta.brace.round.ts"],
      settings: {
        foreground: c.uno4,
      },
    },
    {
      name: "Storage, Type",
      scope: [
        "storage.type.tsx",
        "meta.var.expr.tsx",
        "keyword.control",
        "keyword.control.import.tsx",
        "keyword.control.from.tsx",
        "storage.type.type.tsx",
        "meta.type.declaration.tsx",
        "meta.import.tsx",
      ],
      settings: {
        foreground: c.uno3,
      },
    },

    // I want Debugger to be a different color than the rest of the keywords
    {
      name: "Debugger",
      scope: ["keyword.other.debugger"],
      settings: {
        foreground: c.syntaxAccent,
        fontStyle: "bold",
      },
    },
    {
      name: "Variables",
      scope: [
        "variable",
        "variable.parameter.tsx",
        "variable.object.property.tsx",
        "variable.other.constant.tsx",
        "variable.other.object.tsx",
        "variable.other",
      ],
      settings: {
        foreground: c.duo2,
      },
    },
  ]);

  return theme;
};

const themesDir = join(__dirname, "..", "themes");
// write theme to json file
writeFileSync(
  join(themesDir, "two-firewatch-color-theme-dark.json"),
  JSON.stringify(makeTheme(darkColors), null, 2)
);
writeFileSync(
  join(themesDir, "two-firewatch-color-theme-light.json"),
  JSON.stringify(makeTheme(lightColors), null, 2)
);
